### PR TITLE
Themeapi

### PR DESCRIPTION
**Pull request by pwpatton**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**
Added NG Soho.theme api and exposed it generically and within the soho-personalize.directive.

**Steps necessary to review your pull request (required)**:
- pull branch
- npm i, npm build, npm start
- run http://localhost:4200/ids-enterprise-ng-demo/personalize-color-api
- change colors and themes there, changes should be reflected in the main more menu and vice versa.

### CODE REVIEW

File: projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts [hunk 1]
```diff
@@ -72,28 +71,51 @@ export class SohoPersonalizeDirective implements AfterViewInit, OnDestroy {
...
+   * Return a list of all the available themes.
+   * @returns The list of themes.
+   */
+  public themes(): SohoTheme[] {
+    return Soho.theme.themes();
+  }
+
+  /**
+   * Return the colors used in the current theme that are recommended for personalization.
+   * @returns An object full of the colors with id, name abd hex value.
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts]: is abd supposed to be and?

File: projects/ids-enterprise-ng/src/lib/soho.d.ts [hunk 2]
```diff
@@ -59,5 +60,45 @@ interface SohoBreakpointsStatic {
...
+  name: string;
+}
+
+type SohoPersonalizationColors = { [key: string]: SohoPersonalizationColor };
+
+interface SohoPersonalizationColor {
+  id: string; // id of the color
+  name: string; // human readable name of the color
+  value: string; // hex value of the color
+  backgroundColorClass: string; // class that sets teh background color of an element
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/soho.d.ts]: teh

File: projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts (hunk 1 above)
Comment (pwpatton) [reply projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts]: yes

File: projects/ids-enterprise-ng/src/lib/soho.d.ts [hunk 3]
```diff
@@ -59,5 +60,45 @@ interface SohoBreakpointsStatic {
...
+
+  /**
+   * Return a list of all the available themes.
+   * @returns {array} The list of themes.
+   */
+  themes: () => SohoTheme[];
+
+  /**
+   * Return the colors used in the current theme that are recommended for personalization.
+   * @returns {object} An object full of the colors with id, name abd hex value.
```
Comment (fitzorama) [inline projects/ids-enterprise-ng/src/lib/soho.d.ts]: abd again

File: src/app/application-menu/application-menu.demo.html [hunk 4]
```diff
@@ -169,6 +169,7 @@
   <div class="accordion-header"><a href="javascript:void(0);"><span>P</span></a></div>
   <div class="accordion-pane">
     <div class="accordion-header list-item"><a [routerLink]="['pie']"><span>Pie Chart</span></a></div>
+    <div class="accordion-header list-item"><a [routerLink]="['personalize-color-api']"><span>Personalize Color API</span></a></div>
```
Comment (fitzorama) [inline src/app/application-menu/application-menu.demo.html]: should go up one

File: src/app/personalize-menu/personalize-menu.component.ts [hunk 5]
```diff
@@ -82,18 +69,29 @@ export class PersonalizeMenuComponent implements OnInit {
     const currentTheme = this.personalize.theme = this.theme;
     const currentColour = this.personalize.colors = this.colour;
 
+    this.themeMenuItems = this.personalize.themes();
+
+    const personalizationColors = this.personalize.personalizationColors();
+    this.colourMenuItems = Object.keys(personalizationColors).map(colorId => personalizationColors[colorId]);
+
+    this.setSelectedTheme(currentTheme);
+    this.setSelectedColor(currentColour);
```
Comment (fitzorama) [inline src/app/personalize-menu/personalize-menu.component.ts]: not just this line, but it looks like a lot of variation between American English and British English > colour versus color

Comment (pwpatton) [reply src/app/personalize-menu/personalize-menu.component.ts]: yea, that's how @bthharper build it originally
I think I'll just change back to use American English

File: src/app/personalize/personalize-color-api.demo.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  onColorChange(event: JQuery.TriggeredEvent) {
+    const colorId = event.data;
+    if (colorId === this.currentColorId) {
+      return;
+    }
+    this.currentColorId = colorId;
+
+    const colorHex = this.personalize.personalizationColors()[colorId].value;
+    this.personalize.colors = colorHex;
+    this.ref.markForCheck(); // mark for check so personalization.colors input in messaged
```
Comment (fitzorama) [inline src/app/personalize/personalize-color-api.demo.ts]: comment is used three times, believe it should say 'is messaged' ?

**Comment by bthharper** [comment]
@pwpatton  - do you want this in now, and fix up the typos later?

File: projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts [hunk 7]
```diff
@@ -55,6 +48,12 @@ export class SohoPersonalizeDirective implements AfterViewInit, OnDestroy {
 
   @Output() changecolors = new EventEmitter<SohoChangeColorsPersonalizeEvent>();
 
+  /** EP api */
```
[the remark this answers is not in the document]
Comment (pwpatton) [reply projects/ids-enterprise-ng/src/lib/personalize/soho-personalize.directive.ts]: I tend to use decorated things at the top.
Hostbindings, style and class bindings, inputs, outputs, then internal variables, constructor, then functions.

File: src/app/personalize/personalize-color-api.demo.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  private currentColorId;
+
+  constructor(private ref: ChangeDetectorRef) {}
+
+  ngOnInit() {
+    this.themes = this.personalize.themes();
+    this.colors = this.personalize.personalizationColors();
+
+    this.model.themeId = this.personalize.currentTheme.id;
+    this.model.colorId = 'default';
```
[the remark this answers is not in the document]
Comment (pwpatton) [reply src/app/personalize/personalize-color-api.demo.ts]: yea, we don't really have current color settings from this api. I do listen to `onThemeChanged` and `onColorChanged` and will get those events to set the api demo to whatever menu has set them as. I'd say this is good enough for now.

File: src/app/personalize/personalize-color-api.demo.html [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      </div>
+    </div>
+
+    <div class="row">
+      <div class="field">
+        <label for="color" class="label">Theme Colors</label>
+        <select soho-dropdown soho-busyindicator noSearch id="color" name="color" (change)="onColorChange($event)" [(ngModel)]="model.colorId">
+          <option
+            *ngFor="let color of colors | keyvalue"
+            [attr.data-icon]="JSON.stringify({ icon: 'swatch', class: colors[color.key].backgroundColorClass + ' swatch'})"
```
[the remark this answers is not in the document]
Comment (pwpatton) [reply src/app/personalize/personalize-color-api.demo.html]: I can fix that up.

**Pull request merged**